static/code/explorer.js

```diff
@@ -0,0 +1,623 @@
+ * properties on the object, the second array contains the properties on the
+ * object's prototype, and so on.
+ * @type {!Array<!Array<string>>}
+ */
+Code.Explorer.autocompleteData = [];
+
+/**
+ * Remove any properties that are shadowed by objects higher on the inheritance
+ * chain.  Also sort the properties alphabetically.
+ * @param {Array<!Array<string>>} data Property names from Code City.
+ */
+Code.Explorer.filterShadowed = function(data) {
+  if (!data || data.length < 2) {
+    return;
+  }
+  var properties = Object.create(null);
+  for (var i = 0; i < data.length; i++) {
+    var datum = data[i];
+    for (var j = datum.length - 1; j >= 0; j--) {
+      var prop = datum[j];
+      if (properties[prop]) {
+        datum.splice(j, 1);
+      } else {
+        properties[prop] = true;
+      }
+    }
+    data[i].sort(Code.Explorer.caseInsensitiveComp);
+  }
+};
+
+/**
+ * Comparison function to sort strings A-Z without regard to case.
+ * @param {string} a One string.
+ * @param {string} b Another string.
+ * @return {number} -1/0/1 comparator value.
+ */
+Code.Explorer.caseInsensitiveComp = function(a, b) {
+  a = a.toLowerCase();
+  b = b.toLowerCase();
+  return (a < b) ? -1 : ((a > b) ? 1 : 0);
+};
+
+/**
+ * Don't show any autocompletions if the cursor isn't at the end.
+ */
+Code.Explorer.autocompleteCursorMonitor = function() {
+  var input = document.getElementById('input');
+  if (typeof input.selectionStart === 'number' &&
+      input.selectionStart !== input.value.length) {
+    Code.Explorer.hideAutocompleteMenu();
+    return true;
+  }
+  return false;
+};
+
+/**
+ * Display the autocomplete menu, populated with the provided options.
+ * @param {!Array<string>} options Array of options.
+ * @param {number} index Left offset (in characters) to position menu.
+ */
+Code.Explorer.showAutocompleteMenu = function(options, index) {
+  if (Code.Explorer.autocompleteCursorMonitor()) {
+    return;
+  }
+  var scrollDiv = document.getElementById('autocompleteMenuScroll');
+  scrollDiv.innerHTML = '';
+  for (var i = 0; i < options.length; i++) {
+    var div = document.createElement('div');
+    div.appendChild(document.createTextNode(options[i]));
+    div.addEventListener('mouseover', Code.Explorer.autocompleteMouseOver);
+    div.addEventListener('mouseout', Code.Explorer.autocompleteMouseOut);
+    div.setAttribute('data-option', options[i]);
+    scrollDiv.appendChild(div);
+  }
+  var menuDiv = document.getElementById('autocompleteMenu');
+  menuDiv.style.display = 'block';
+  menuDiv.scrollTop = 0;
+  var left = Math.round(index * Code.Explorer.SIZE_OF_INPUT_CHARS);
+  var maxLeft = window.innerWidth - menuDiv.offsetWidth;
+  menuDiv.style.left = Math.min(left, maxLeft) + 'px';
+};
+
+/**
+ * Stop displaying the autocomplete menu.
+ */
+Code.Explorer.hideAutocompleteMenu = function() {
+  document.getElementById('autocompleteMenu').style.display = 'none';
+  Code.Explorer.autocompleteSelect(null);
```

Comment:
When we look to see if an option is selected (e.g. when 'enter' is pressed), we would otherwise find an option, even if it isn't visible.  Deselecting it here prevents having to check for the visibility of the menu in several other places.